### DISCUSSION TITLE
SQL Queries as ORM columns/attributes

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#7975] [Usage Questions] [answered]
Hey fantastic community,

Please help me make this elegant. Here is my fictitious data situation:

<img width="165" alt="CleanShot 2022-04-29 at 16 45 16@2x" src="https://user-images.githubusercontent.com/1538854/166066862-33b22d4a-8c8b-4753-96ef-d4504d6767d3.png">

A company has multiple budgets.
Each budget contains several transactions.
And each transactions can have multiple bank transactions attached to it.

Each one of those objects are SQLAlchemy Model. I am trying to build summary fields on higher objects.

i.e: When I pull my `/company/2` endpoint, I want to include the Company object but also the sum of all BankAccountTransactions linked to it. Or I want to display the sum of a field on the Transaction object.

I am trying to do this in an elegant fashion on the Company object, Im kinda looking for this:

```
total_cost = <SQL Query to get exactly the data I want to return here>[0]
```

I want to do this in a way that I can avoid having to load all child relationships down to the bottom to calculate those on the fly, instead I want a specific sub-query to run when SQLAlchemy go fetches the Company object and call it like I would for any other columns on that model.

I am lost between hybrid_attribute, column_property, etc...

How would you guys do this?

Phil

**Comment by zzzeek** [comment, marked as answer]
SQLAlchemy has moved away from the pattern of `query_object[<index>]` that automatically invokes a query as a best practice.   You certainly can still do that kind of thing, and the dynamic relationship strategy still does that as well as the `Query` object, but it's not what *I* would do.

but either way the way I approach this kind of problem is start by writing some of those "exact query that I want" objects and see what those look like.   it looks like you want aggregates in these, so indeed these would be SQL select() objects that you'd get through whatever means you'd like, then execute() them in a Session to get a result.   

if the issue is how to write Query / select() with aggregates using relatoinships to create the joins, I can show you that if you want to request a specific example.   The general idea is illustrated at https://docs.sqlalchemy.org/en/14/orm/queryguide.html#joins .

overall if these special queries involve an independent "select" object that has things like aggregates in it, this usually indicates that it's not any kind of "hybrid" / "synonym" kind of thing, just use regular Python methods.
